### PR TITLE
Fixes Shell Side Arm

### LINKED ISSUE #1868: Shell Side Arm not implemented correctly
Sorry, I might be a bit late to this, but I happened to be going through some of these pull requests and noticed there are a few problems here.

1. The way Shell Side Arm was implemented is inconsistent with how it actually works in the official games. It's supposed to choose the damage category at the beginning of the turn, and not mid-turn. See [here](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/post-8508168).
2. It doesn't take into account Reflect or Light Screen. See link in 1.
3. It doesn't take into account Wonder Room. See [here](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/post-8707046). Similar things like Body Press and Download may be implemented incorrectly in other places in your battle engine, but I haven't checked.
4. The AI would have no idea how to use it properly due to this implementation. If you hooked it directly into the damage calculator, it would work much better for the AI, and you wouldn't have to worry about a global variable being set incorrectly.
5. In Link Battles the animation wouldn't play correctly because ``gSwapDamageCategory`` isn't being sent over.

This is also just a general note based on things I've seen in a few places, but is there a specific reason global variables are being used to handle things like this? Why not just add a field in ``gBattleStruct`` instead? It's automatically cleared at the beginning of each battle, and it takes up no extra static memory.

_Originally posted by @Skeli789 in https://github.com/rh-hideout/pokeemerald-expansion/issues/1817#issuecomment-962477674_

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Fixes  #1868

What a waste of cycles and nobody besides Skeli noticed 😅 (not that we shouldn't fix it)

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
I can't figure out why the AI tests fail though it is also wrong ingame

**Comment by AlexOn1ine** [comment]
> I can't figure out why the AI tests fail though it is also wrong ingame. What is weird about this is that we have plenty of similar AI tests that all pass

ready for review

File: src/battle_ai_util.c [hunk 1]
```diff
@@ -477,8 +478,9 @@ s32 AI_CalcDamage(u32 move, u32 battlerAtk, u32 battlerDef, u8 *typeEffectivenes
     }
     else if (gMovesInfo[move].effect == EFFECT_PHOTON_GEYSER)
         gBattleStruct->swapDamageCategory = (GetCategoryBasedOnStats(gBattlerAttacker) == DAMAGE_CATEGORY_PHYSICAL);
-
-    if (gMovesInfo[move].effect == EFFECT_NATURE_POWER)
+    else if (move== MOVE_SHELL_SIDE_ARM && gBattleStruct->shellSideArmCategory[battlerAtk][battlerDef])
```
Comment (Skeli789) [inline src/battle_ai_util.c:481]:
```suggestion
    else if (move == MOVE_SHELL_SIDE_ARM && gBattleStruct->shellSideArmCategory[battlerAtk][battlerDef])
```

File: src/battle_util.c [hunk 2]
```diff
@@ -8304,15 +8304,15 @@ bool32 IsBattlerProtected(u32 battler, u32 move)
     // Protective Pads doesn't stop Unseen Fist from bypassing Protect effects, so IsMoveMakingContact() isn't used here.
     // This means extra logic is needed to handle Shell Side Arm.
     if (GetBattlerAbility(gBattlerAttacker) == ABILITY_UNSEEN_FIST
-        && (gMovesInfo[move].makesContact || (gMovesInfo[move].effect == EFFECT_SHELL_SIDE_ARM && gBattleStruct->swapDamageCategory))
-        && !gProtectStructs[battler].maxGuarded) // Max Guard cannot be bypassed by Unseen Fist
+     && (gMovesInfo[move].makesContact || (move == MOVE_SHELL_SIDE_ARM && gBattleStruct->shellSideArmCategory[battler][gBattlerTarget]))
```
Comment (Skeli789) [inline src/battle_util.c:8307]:
```suggestion
     && (gMovesInfo[move].makesContact || (move == MOVE_SHELL_SIDE_ARM && gBattleStruct->shellSideArmCategory[gBattlerAttacker][battler]))
```

Comment (Skeli789) [reply src/battle_util.c:8307]: Better to pass the attacker into this function itself but whatever

File: src/battle_util.c [hunk 3]
```diff
@@ -11382,3 +11382,53 @@ void RemoveBattlerType(u32 battler, u8 type)
...
+            targetSpDefStat = gBattleMons[battlerDef].spDefense;
+            statStage = gBattleMons[battlerDef].statStages[STAT_SPDEF];
+            targetSpDefStat *= gStatStageRatios[statStage][0];
+            targetSpDefStat /= gStatStageRatios[statStage][1];
+
+            special = ((((2 * gBattleMons[battlerAtk].level / 5 + 2) * gMovesInfo[MOVE_SHELL_SIDE_ARM].power * attackerSpAtkStat) / targetSpDefStat) / 50);
+
+            if (((physical > special) || (physical == special && RandomPercentage(RNG_SHELL_SIDE_ARM, 50))))
+            {
+                gBattleStruct->shellSideArmCategory[battlerAtk][battlerDef] = TRUE;
```
Comment (Skeli789) [inline src/battle_util.c:11430]: Instead of setting a boolean, it would be clearer if you set the category itself to this field.

**Review by Skeli789 — CHANGES_REQUESTED** [review]

File: src/battle_util.c (hunk 2 above)
Comment (AlexOn1ine) [reply src/battle_util.c:8307]: I agree. Will change

**Comment by AlexOn1ine** [comment]
applied reviews.

**Review by Bassoonian — APPROVED** [review]

File: src/battle_util.c (hunk 2 above)
Comment (Skeli789) [reply src/battle_util.c]: Probably `IsMoveMakingContact` should be changed as well

Comment (AlexOn1ine) [reply src/battle_util.c]: Will do. Thx for the follow up.

**Pull request merged**